Admin/BreadcrumbsBuilder.php

```diff
@@ -0,0 +1,124 @@
+
+use Knp\Menu\ItemInterface as MenuItemInterface;
+
+final class BreadcrumbsBuilder implements BreadcrumbsBuilderInterface
+{
+    /**
+     * Get breadcrumbs for $action.
+     *
+     * @param Admin  $admin
+     * @param string $action
+     *
+     * @return ItemInterface
+     */
+    public function getBreadcrumbs($admin, $action)
+    {
+        $breadcrumbs = array();
+        if ($admin->isChild()) {
+            return $admin->getParent()->getBreadcrumbs($action);
+        }
+
+        $menu = $admin->buildBreadcrumbs($action);
+
+        do {
+            $breadcrumbs[] = $menu;
+        } while ($menu = $menu->getParent());
+
+        $breadcrumbs = array_reverse($breadcrumbs);
+        array_shift($breadcrumbs);
+
+        return $breadcrumbs;
+    }
+
+    /**
+     * Generates the breadcrumbs array.
+     *
+     * Note: the method will be called by the top admin instance (parent => child)
+     *
+     * @param Admin              $admin
+     * @param string             $action
+     * @param ItemInterface|null $menu
+     *
+     * @return ItemInterface
+     */
+    public function buildBreadcrumbs($admin, $action, MenuItemInterface $menu = null)
+    {
+        if (!$menu) {
+            $menu = $admin->getMenuFactory()->createItem('root');
+
+            $menu = $menu->addChild(
+                $admin->trans(
+                    $admin->getLabelTranslatorStrategy()->getLabel(
+                        'dashboard',
+                        'breadcrumb',
+                        'link'
+                    ),
+                    array(),
+                    'SonataAdminBundle'
+                ),
+                array('uri' => $admin->getRouteGenerator()->generate(
+                    'sonata_admin_dashboard'
+                ))
+            );
+        }
+
+        $menu = $menu->addChild(
+            $admin->trans(
+                $admin->getLabelTranslatorStrategy()->getLabel(sprintf(
+                    '%s_list',
+                    $admin->getClassnameLabel()
+                ), 'breadcrumb', 'link')
+            ),
+            array(
+                'uri' => $admin->hasRoute('list') && $admin->isGranted('LIST') ?
+                $admin->generateUrl('list') :
+                null,
+            )
+        );
+
+        $childAdmin = $admin->getCurrentChildAdmin();
+
+        if ($childAdmin) {
+            $id = $admin->getRequest()->get($admin->getIdParameter());
+
+            $menu = $menu->addChild(
+                $admin->toString($admin->getSubject()),
+                array('uri' => $admin->hasRoute('edit') && $admin->isGranted('EDIT') ?
+                    $admin->generateUrl('edit', array('id' => $id)) :
+                    null, )
+            );
+
+            return $childAdmin->buildBreadcrumbs($action, $menu);
+        }
+
+        if ($action === 'list' && $admin->isChild()) {
+            $menu->setUri(false);
+        } elseif ($action !== 'create' && $admin->hasSubject()) {
```

Comment:
> It's on Symfony standards.

https://github.com/sonata-project/SonataAdminBundle/pull/3791#discussion_r62795450

From http://symfony.com/doc/current/contributing/code/standards.html#structure:

> Use Yoda conditions when checking a variable against an expression to avoid an accidental assignment inside the condition statement (this applies to `==`, `!=`, `===`, and `!==`);

I'm not saying your arguments is acceptable or not. But if we do not follow a standard, we would have Yoda and non-yoda conditions everywhere (this is already the case).

As 99% of sonata projects are bundle, this makes sense to follow Symfony standards that require yoda conditions for these operators.